rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -31,57 +31,75 @@ Status        | Proposed
+TFX component that transforms the training data by invoking TFT and
+collects pre-transform and post-transform statistics of the data by invoking
+TFDV.
+
+Currently, each TFX library accepts different in-memory data representation:
 
 |    | TFDV | TFT  | TFMA | BulkInference |
 |:---| :--- | :--- | :--- | :------------ |
 | In-memory data representation | Arrow RecordBatches | Dict[str, np.ndarray] | str (raw data records), Dict[str, np.ndarray] | str (raw data records) |
 | Understand the data and conduct analysis | input data is encoded losslessly as RecordBatches. | the in-mem representation may be lossy. | Relies on the model’s input layer, and the format is Dict[str, np.ndarray]. | N/A |
 | Feed TF | N/A | the in-mem representation is TF feedable. | Feed “raw data” to the model. | Feed “raw data” to the model |
 
-This has created many issues:
-
-*   Users of individual components need to adapt their data (if input format not
-    already supported) to each component they want to use.
-*   Individual components rely on unenforceable assumptions on how to interpret
-    the input data consistently.
-*   The complexity of adding new logical data representations (for example,
-    tf.SequenceExample) scales with the number of components.
+When a TFX component needs to invoke multiple TFX libraries, it may need to
+decode the data into one of the libraries’ in-memory representations,
+and translate that into the other library’s. For example, in the “Transform”
+component mentioned above:
+
+![alt_text](20191017-tfx-standardized-inputs/double-translation.png)
+
+Note that TFDV is invoked twice, with different data. Thus the translation needs
+to happen twice.
+
+This has created several issues:
+
+*   The translation is computationally expensive. In a real world set-up, such
+    translation could take as many CPU cycles as the core TFT and TFDV logic
+    takes in total.
+*   More such translation logic may need to be implemented to support expanding
+    TFX use cases -- imagine that TFMA needs to invoke TFDV to compute
+    statistics over slices of data identified by model evaluation.
+*   The complexity of adding new logical data representations scales with the
+    number of components. For example, to add SequenceExample support in TFX,
+    one may need to come up with an in-memory representation for each of the
+    components, to keep the consistency within the component.
+*   TFX library users whose data format is not supported natively by TFX would
+    have to implement the decoding logic for each of the libraries
+    they want to use.  For example, had TFX not supported the CSV format, a user
+    would have to implement
+    [one CSV decoder for TFT](https://github.com/tensorflow/transform/blob/master/tensorflow_transform/coders/csv_coder.py),
+    and [another CSV decoder for TFDV](https://github.com/tensorflow/data-validation/blob/master/tensorflow_data_validation/coders/csv_decoder.py)
```

Comment:
Why is this the case? In TF you can implement a parsing op once and use it many times; why do TFX components care so much about the format the data is in?

As in, why do we care about the format of the data at rest instead of just assuming that there is a tf function or whatever that will take an at-rest format into the arrays you expect?

Reply:
You are absolutely right, but you are making the assumption that everybody agrees on the same in-memory format and this is not true in TFX (libraries) currently, as described in this section. Please note we are covering a portfolio of TFX libraries that emerged independently, and have evolved relatively independently so far.